### DISCUSSION TITLE
struggling with Reactflow initialize->measure->layout->render sequence

### DISCUSSION THREAD

**Discussion by samwan-loops** [xyflow/xyflow#2973] [General]
Hello, I've been struggling with the rendering sequence for ReactFlow, as well as adapting my workarounds through successive iterations of this library. I was wondering if you could advice on best practice with the latest v11 release.

We have a dynamic data source, and when the data changes, I generate a new set of Node[] with the custom data, and use a custom node component to render that data. The data results in non-uniform node dimensions. I've written a tree-based layout algorithm that calculates the positioning for the nodes with non-uniform node dimensions according to our specific needs.

I've run into a few issues with reliable initialization and node dimension measurement. The main struggling through different versions of ReactFlow is that the nodes will render in their default positions BEFORE the node dimensions can be measured, before the layout positions can be calculated, and before the positions can be applied to the nodes. I've done various workarounds, such as setting the opacity to 0 at first, or setting position to (-10000,-10000), but it's been difficult to prevent flickering for the first render cycle without somewhat convoluted code.

Other struggles have included...
* figuring out whether the external nodes or the internal state nodes should be used for measurement and position assignment
* figuring out how to add animation transitions and trigger the renderer to update edges during the transitions
* figuring out when ` const nodesInitialized = useStore(nodesInitializedSelector);` is accurate vs when to manually test for any undefined node widths or heights.

Can you provide documentation or sample code that describes best practices for this kind of customization? Specifically something that will point out how to measure, calculate layout, and apply layout for an array of non-uniform nodes that is frequently recreated from an external data source?

Much obliged!
--Sam

**Comment by chrtze** [comment]
Hi Sam,

I understand that there can be some confusion around the render/update order especially when working with a custom layout algorithm. Internally, once a node gets added, we attach a `ResizeObserver` to it to update the dimensions in the React Flow state. Because the `ResizeObserver` needs a DOM node, we need to render the nodes first and update the dimensions of the nodes in the state after that.

That's why it needs some workaround to prevent flickering if you want to layout the nodes using the dynamic dimensions from the state. A possible solution for this has been implemented in the [Dagre Auto Layout example](https://pro.reactflow.dev/examples/auto-layout-dagre). As the layout algorithm takes the node dimensions into account, we wait for all nodes to be initialized with their dimensions before running the layout for the first time. If a node is added dynamically, we add it with `opacity: 0` and after the layout is run, it is set to `opacity: 1`. Did you follow that examples for your implementation?

Of course, the easiest and most reliable way would be to calculate the dimensions for your nodes based on their content so that you can run the layout algorithm before you pass the nodes to the React Flow component with their final positions.

> figuring out whether the external nodes or the internal state nodes should be used for measurement and position assignment

Not sure if I got this right but in the layout examples, we create a trigger from the store that tells us when the layout algorithm should run (for example: `nodesInitialized` or `nodeCount`). Whenever this trigger changes, the layout algorithm gets the current nodes from the internal state by using the `getNodes` method from the React Flow instance. This makes sure that we always have the state of the nodes that are currently being displayed.

> figuring out how to add animation transitions and trigger the renderer to update edges during the transitions

For an example of transition nodes between two positions, you can see the [Expand and Collapse example](https://pro.reactflow.dev/examples/expand-collapse). This example uses a layout algorithm, too, so you can see how the two functions can work together. If you update a node position, the connected edges will be animated automatically. Using CSS transitions for the positions is not practical here because of the edges, although it would be better for the performance.

> figuring out when  const nodesInitialized = useStore(nodesInitializedSelector); is accurate vs when to manually test for any undefined node widths or heights.

I think this depends on your use case. Could you provide some more information or example?

**Reply by samwan-loops** [reply]
Thank you very much for this thorough explanation. I spent last week rewriting our implementation to use the approach in the Dagre Auto Layout example. However, new nodes still flicker temporarily in Safari (but not Chrome) before the layout can be applied. The Dagre Auto Layout example does not flicker in Safari. I will look into this further and might try rebuilding the flowchart again from scratch with your explanation of the sequence in mind, to see if I've missed something.
 
Can you confirm that the `ReactFlow instance getNodes()` will always be a mirror of the `nodes={nodes}` component prop, only with assigned position and dimensions? Our code is a bit too involved to paste online as an example, but I think `nodesInitialized` is sometimes not correct (some internal nodes have undefined width and height) when adding a node dynamically. I will think of how to write a paradigmatic example of what we're trying to do, to see if it can isolate the problem or at least explain the use-case.

Regarding the animated transitions, are you relying on React's repeated render cycle to step through the animation? If my project requires dynamic layout, dynamic sets of nodes, and non-uniform node dimensions, would you pipe the array of nodes through the following transformations?

original nodes -> internal measured nodes -> layouted nodes -> animated nodes?

**Reply by chrtze** [reply]
> However, new nodes still flicker temporarily in Safari (but not Chrome) before the layout can be applied

Interesting. Maybe it would make sense for us to build something like an auto layout template where you can adjust the layout algorithm. Also, I understand that your use case is slightly different from the Auto Layout example as you are replacing parts of the existing data (updating existing nodes) and not only adding new nodes.

> Can you confirm that the ReactFlow instance getNodes() will always be a mirror of the nodes={nodes} component prop, only with assigned position and dimensions?

Yes.

> some internal nodes have undefined width and height when adding a node dynamically

Yes, this is intended. The node gets added to the state with undefined width and height before the `ResizeObserver` can be added and update the dimensions in the internal state.

> Regarding the animated transitions, are you relying on React's repeated render cycle to step through the animation? If my project requires dynamic layout, dynamic sets of nodes, and non-uniform node dimensions, would you pipe the array of nodes through the following transformations?

> original nodes -> internal measured nodes -> layouted nodes -> animated nodes?

Yes. The closest example for this kind of functionality would be the [Expand and Collapse example](https://pro.reactflow.dev/examples/expand-collapse) in combination with the [Dagre Auto Layout example](https://pro.reactflow.dev/examples/auto-layout-dagre) which supports dynamic node dimensions.

Let me know if you need further help.

**Reply by shadiramadan** [reply]
> something like an auto layout template where you can adjust the layout algorithm

That would be really nice- I'm having some of the same issues fighting the layout system w/ dynamic content.
YAML (editor) -> XYFlow Nodes/Edges

**Reply by lucasgelfond** [reply]
@chrtze I have a bit more complex of a case, that I really cannot manage to figure out without flickering! 

I have successfully implemented a single: invisibly mount nodes to graph to get their size, measure them, do a second layout pass with Dagre to get them to stay. 


More complicated, however, is that I have a parent node whose dimensions are based on the layout of its children.

This has been a (rough) sketch of the algorithm I am using, happy also to provide code / reproducible example, etc if helpful:
- first, mount ONLY sub-nodes
- layout sub-nodes nodes - I cannot do this unless they are visible
- from laid out graph, get size of the root-level node that is their parent; change the corresponding root-level node's size
- use `setNodes` to mount all root and sub-level nodes to get their size
- lay them out using dagre, but adjust the positions of all of the sub-nodes by the position of their parent so that they are placed inside

I can't come up with a way to do this that doesn't involve a visible layout pass. Also, it changes the viewport in a really jerky manner because of `reactFlow.setCenter`. 

Any ideas? I avoided Elkjs because of bundle size limitations, seems like this is the correct pattern with Dagre/React -Flow though? Just feels quite complex / clunky, etc, thinking there may be a better way to do?

**Comment by DylanMerigaud** [comment]
I fought this exact rendering-sequence flicker with non-uniform node sizes. What fixed it: seed the nodes with `visibility: hidden`, wire `onNodesChange` so React Flow's measurements flow back, gate the layout on `useNodesInitialized()`, then flip to `visibility: visible` in the same pass that applies positions. The detail that mattered was reading sizes off the `dimensions` change events (never stale) instead of re-reading `node.measured` in an effect, which left my handles a few px off for a tick.

I ended up packaging that whole render-hidden -> measure -> layout -> reveal dance into a hook so I stopped rewriting it, in case it's useful: https://www.npmjs.com/package/react-flow-auto-layout (dagre under the hood, re-runs on size changes, live demo in the readme). The source is small if you'd rather just lift the pattern.
